File: hosts/libretro/button_repeat.cpp
#include "button_repeat.h"

namespace bmsx {
namespace {
constexpr i32 INITIAL_REPEAT_DELAY_FRAMES = 15;
constexpr i32 REPEAT_INTERVAL_FRAMES = 4;
}

void ButtonRepeat::reset() {
	active = false;
	nextRepeatAtMs = 0;
	lastFrameEvaluated = -1;
	lastResult = false;
}
bool ButtonRepeat::update(bool pressed, bool justPressed, f64 pressedAtMs, f64 now, f64 frameDurationMs, i64 frameId) {
	if (lastFrameEvaluated == frameId) return lastResult;
	lastFrameEvaluated = frameId;
	lastResult = justPressed;
	if (!pressed) {
		active = false;
		nextRepeatAtMs = 0;
	} else {
		if (justPressed || !active) {
			active = true;
			nextRepeatAtMs = pressedAtMs + INITIAL_REPEAT_DELAY_FRAMES * frameDurationMs;
		}
		if (!justPressed && now >= nextRepeatAtMs) {
			nextRepeatAtMs += REPEAT_INTERVAL_FRAMES * frameDurationMs;
			lastResult = true;
		}
	}
	return lastResult;
}
} // namespace bmsx

File: hosts/common/input/button_repeat.ts
const INITIAL_REPEAT_DELAY_FRAMES = 15;
const REPEAT_INTERVAL_FRAMES = 4;

/** Host-time repeat cadence shared by UI input and per-player IDE queries. */
export class ButtonRepeat {
	private active = false;
	private nextRepeatAtMs = 0;
	private lastFrameEvaluated = -1;
	private lastResult = false;

	public reset(): void {
		this.active = false;
		this.nextRepeatAtMs = 0;
		this.lastFrameEvaluated = -1;
		this.lastResult = false;
	}

	public update(pressed: boolean, justPressed: boolean, pressedAtMs: number, now: number, frameDurationMs: number, frameId: number): boolean {
		if (this.lastFrameEvaluated === frameId) return this.lastResult;
		this.lastFrameEvaluated = frameId;
		this.lastResult = justPressed;
		if (!pressed) {
			this.active = false;
			this.nextRepeatAtMs = 0;
		} else {
			if (justPressed || !this.active) {
				this.active = true;
				this.nextRepeatAtMs = pressedAtMs + INITIAL_REPEAT_DELAY_FRAMES * frameDurationMs;
			}
			if (!justPressed && now >= this.nextRepeatAtMs) {
				this.nextRepeatAtMs += REPEAT_INTERVAL_FRAMES * frameDurationMs;
				this.lastResult = true;
			}
		}
		return this.lastResult;
	}
}
